pkg/reporterreporter/report.go

```diff
@@ -30,49 +39,137 @@ type RequestData struct {
 	Data      reporter_v1.NamespaceReport `json:"data"`
 }
 
-func (r *Report) Run(ctx context.Context) error {
-	namespaces, err := r.KubernetesClient.CoreV1().Namespaces().List(meta_v1.ListOptions{})
+func NewReport(slurperURI string, cluster string, reporterClient client.Interface, kubeClient kubernetes.Interface, logger *zap.Logger) ctrl.Server {
+	logger.Info("ReporterReporter configured",
+		zap.String("cluster_id", cluster),
+		zap.String("slurper_uri", slurperURI),
+		zap.String("slurper_path", slurperPath))
+
+	return &report{
+		clusterID:        cluster,
+		httpClient:       util.HTTPClient(),
+		kubernetesClient: kubeClient,
+		logger:           logger,
+		mutator: restclient.NewRequestMutator(
+			restclient.BaseURL(slurperURI),
+			restclient.JoinPath(slurperPath),
+			restclient.Method(http.MethodPost)),
+		reporterClient: reporterClient,
+	}
+}
+
+func isRetriableError(statusCode int) bool {
+	_, ok := map[int]bool{
+		http.StatusRequestTimeout:  true,
+		http.StatusTooManyRequests: true,
+	}[statusCode]
+
+	if ok {
+		return true
+	}
+
+	return statusCode >= 500 && statusCode <= 599
+}
+
+// sendData will send the requestData containing the report to slurper
+func (r *report) sendData(requestData RequestData) (retriable bool, err error) {
+	req, err := r.mutator.NewRequest(restclient.BodyFromJSON(requestData))
 	if err != nil {
-		return errors.Wrap(err, "failed to list namespaces")
+		return false, errors.Wrap(err, "unable to craft a HTTP request for slurper")
 	}
-	for _, namespace := range namespaces.Items {
-		list, err := r.ReporterClient.ReporterV1().Reports(namespace.Name).List(meta_v1.ListOptions{})
-		if err != nil {
-			return errors.Wrap(err, "could not list reports")
+
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		if urlError, ok := err.(*url.Error); ok {
+			if urlError.Timeout() || urlError.Temporary() {
+				return true, errors.Wrap(err, "transient error occurred during request to slurper")
+			}
 		}
-		if len(list.Items) == 0 {
-			r.Logger.Sugar().Infof("%q is empty", namespace.Name)
-			continue
+
+		return false, errors.Wrap(err, "error occured during request to slurper")
+
+	}
+	defer resp.Body.Close()
```

Comment:
should be on line 82

Reply:
https://sourcegraph.com/github.com/golang/go/-/blob/src/net/http/client.go#L480-484